### PR TITLE
Acknowledge __all__

### LINKED ISSUE #1606: MyPy not acknowledging __all__
When MyPy encounters star (*) imports it considers every symbol defined in the given module as being imported even if an `__all__` attribute is present on the module.

For example, say you have mypkg with two files, main.py:

```python
from datetime import datetime
from mypkg.widgets import *
```

and widget.py:
```python
from datetime import datetime

__all__ = [
    "Gadget"
]

class Gadget(object):
    pass
```

Running "mypy mypkg/app.py" issues the following error: "mypkg/app.py:2: error: Name 'datetime' already defined", though datetime should not be re-imported to the app module.

### PR DESCRIPTION
**Pull request by kirbyfan64**
Fixes #1606. Took me a bit of experimentation to figure out where to put the code...

### CODE REVIEW

**Comment by gvanrossum** [comment]
You have some CI failures; "mypy mypy" fails and a few tests also fail (perhaps as fallout).

File: mypy/semanal.py [hunk 1]
```diff
@@ -159,6 +159,8 @@ class SemanticAnalyzer(NodeVisitor):
     modules = None  # type: Dict[str, MypyFile]
     # Global name space for current module
     globals = None  # type: SymbolTable
+    # Globals that were mentioned in __all__ and should be module_public.
+    global_exports = set()
```
Comment (gvanrossum) [inline mypy/semanal.py]: This variable is unused. Maybe you changed your mind and named it all_exports later?

Comment (gvanrossum) [inline mypy/semanal.py]: Also a mutable default is a bad idea.

File: mypy/semanal.py [hunk 2]
```diff
@@ -250,6 +253,11 @@ def visit_file(self, file_node: MypyFile, fnam: str) -> None:
 
         self.errors.set_ignored_lines(set())
 
+        if '__all__' in self.globals:
+            for name, g in self.globals.items():
+                if name not in self.all_exports:
+                    g.module_public = False
```
Comment (gvanrossum) [inline mypy/semanal.py]: Nice!

File: mypy/semanal.py [hunk 3]
```diff
@@ -1069,6 +1077,11 @@ def visit_assignment_stmt(self, s: AssignmentStmt) -> None:
         self.process_typevar_declaration(s)
         self.process_namedtuple_definition(s)
 
+        if len(s.lvalues) == 1 and isinstance(s.lvalues[0], NameExpr) and \
```
Comment (gvanrossum) [inline mypy/semanal.py]: Use () instead of \ to break long lines.

**Comment by kirbyfan64** [comment]
@gvanrossum Fixed!

**Comment by gvanrossum** [comment]
Thanks!

**Pull request merged**